Sources/MapboxNavigation/NavigationMapView+Annotations.swift

```diff
@@ -0,0 +1,488 @@
+        }
+        
+        if style.image(withId: ImageIdentifier.railroadCrossingNight) == nil,
+           let railroadCrossingNight =  Bundle.mapboxNavigation.image(named: "RailroadCrossingNight") {
+            try style.addImage(railroadCrossingNight, id: ImageIdentifier.railroadCrossingNight)
+        }
+    }
+    
+    func updateIntersectionSignals(with routeProgress: RouteProgress) {
+        guard !routeProgress.routeIsComplete else {
+            removeIntersectionSignals()
+            return
+        }
+        var featureCollection = FeatureCollection(features: [])
+        
+        let stepProgress = routeProgress.currentLegProgress.currentStepProgress
+        let intersectionIndex = stepProgress.intersectionIndex
+        let stepIntersections = stepProgress.intersectionsIncludingUpcomingManeuverIntersection
+        
+        for intersection in stepIntersections?.suffix(from: intersectionIndex) ?? [] {
+            if let feature = signalFeature(from: intersection, styleType: styleType) {
+                featureCollection.features.append(feature)
+            }
+        }
+        
+        let style = mapView.mapboxMap.style
+        
+        do {
+            let sourceIdentifier = NavigationMapView.SourceIdentifier.intersectionSignalSource
+            if style.sourceExists(withId: sourceIdentifier) {
+                try style.updateGeoJSONSource(withId: sourceIdentifier, geoJSON: .featureCollection(featureCollection))
+            } else {
+                var source = GeoJSONSource()
+                source.data = .featureCollection(featureCollection)
+                try style.addSource(source, id: sourceIdentifier)
+            }
+            
+            let layerIdentifier = NavigationMapView.LayerIdentifier.intersectionSignalLayer
+            guard !style.layerExists(withId: layerIdentifier) else { return }
+            
+            var shapeLayer = SymbolLayer(id: layerIdentifier)
+            shapeLayer.source = sourceIdentifier
+            shapeLayer.iconAllowOverlap = .constant(false)
+            shapeLayer.iconImage = .expression(Exp(.get) {
+                "imageName"
+            })
+            
+            let layerPosition = layerPosition(for: layerIdentifier)
+            try style.addPersistentLayer(shapeLayer, layerPosition: layerPosition)
+        } catch {
+            Log.error("Failed to perform operation while adding intersection signals with error: \(error.localizedDescription).",
+                      category: .navigationUI)
+        }
+    }
+    
+    private func signalFeature(from intersection: Intersection, styleType: StyleType) -> Feature? {
+        var properties: JSONObject? = nil
+        let styleTypeString = (styleType == .day) ? "Day" : "Night"
+        if intersection.railroadCrossing == true {
+            properties = ["imageName": .string("RailroadCrossing" + styleTypeString)]
```

Comment:
When possible, feature properties should be semantic rather than presentational, to ensure reusability. Let’s set a `trafficControlDevice` property to `traffic_signals` or `level_crossing`. This would be consistent with the Mapbox Streets source’s `road` layer’s [`class`](https://docs.mapbox.com/data/tilesets/reference/mapbox-streets-v8/#--road---class-text) property.

Reply:
The name of the image identifiers are updated.